### ISSUE TITLE
[GoogleTTS] Google TTS losing auth every few days

### ISSUE THREAD

**Issue by hundertvolt** [openhab/openhab-addons#17425]
I am using Google TTS (with prepaid account) for about half a year.

Ever since and throughout all OH updates I had the issue that the add-on is generally functional after configuration, but stops working with invalid credentials after two to five days.

In the logs, only this one line is issued in the non-functional state:

`2024-09-16 15:11:34.224 [WARN ] [core.voice.internal.VoiceManagerImpl] - Error saying 'Test': No TTS service can be found for voice googletts:deDEStandardE`

The short-term solution is visiting my Google account, request a new auth code (and copy it from the address bar as described in the docs), save it in the GoogleTTS config and have it immediately working again. But only for a few days until this repeats.

I have seen many similar reports in the OpenHAB Forum, while all of them ended either in frustration or in using another TTS service or creating workarounds like a dedicated Home Assistant instance only for bridging OpenHAB to Google TTS. None came up with a stable solution for this behavior.

My most recent [thread](https://community.openhab.org/t/google-tts-losing-auth-every-few-days/158324) did not yet yield any results.

## Expected Behavior
Generate the Authorization code once manually, paste it into the binding, and renewal will from there on happen automatically.

## Current Behavior
The Authorization codebecomes invalid after a strongly variable timespan. It may work for one week, it may fail after one day, while in most cases it stops working after 3 to 4 days. Following the steps for generating an Auth key and replacing it in the binding makes it immediately work for some days again. Note that only the Auth key needs to be renewed; Client ID and Client Secret remain unchanged.

## Possible Solution
Fix automatic key renewal functions

## Steps to Reproduce (for Bugs)
Follow the steps in the docs to create a working GoogleTTS instance, let it run on, try once per day if it still works until no voice output comes, but the error line in the log.

## Context
Voice reminder for several smart home appliances fail allover again, while using the GoogleTTS voice still has by far the best quality.

## Your Environment
OpenHABian 4.2.1 / Raspbian Bullseye on Raspberry Pi 4 / 8 GB
But this already happened from OpenHABian 4.0.x on throughout all updates and versions.

**Comment by lsiepel** [comment]
Looking at the code it makes use of the `OAuthClientService`. Nowhere do i see a check if the token is expired and act on it (call `OAuthClientService.refreshToken()`.
I'm not very familiair with this binding or with the clientservice, but i might look at it in closer detail. Could you provide a debug log ?

**Comment by hundertvolt** [comment]
I'll create a debug level log as soon as the binding fails again.
While I wrote the initial ticket, I created a new auth code to ensure the issue still exists as described.
It does, but now we'll have to wait a bit for the current auth key to expire ;-)

**Comment by lsiepel** [comment]
In the mean time, based on some assumptions, i have created a possible fix with a watch dog that looks after these tokens. When you have a debug log, we can validate and move forward to a fix.

**Comment by hundertvolt** [comment]
Well, here we go. Token from last week expired.
I set the log mode to TRACE and found this (and only this, nothing else so far):

`[WARN ] [ce.googletts.internal.GoogleCloudAPI] - Error authenticating Google Cloud TTS service: Error fetching access token. Invalid authcode? Please generate a new one.`

`[ERROR] [oauth2client.internal.OAuthConnector] - grant type refresh_token to URL https://accounts.google.com/o/oauth2/token failed with error code invalid_grant, description Token has been expired or revoked.`

`[WARN ] [ce.googletts.internal.GoogleCloudAPI] - Error initializing Google Cloud TTS service: Error fetching access token. Invalid authcode? Please generate a new one.`

`[DEBUG] [.googletts.internal.GoogleTTSService] - Synthesize 'Test' for voice 'googletts:deDEStandardE' in format AudioFormat [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100, channels=1]`

`[WARN ] [core.voice.internal.VoiceManagerImpl] - Error saying 'Test': Missing service configuration.`

I hope this gives you some useful information! Of course, I would volunteer for testing your fixes :-)

**Comment by hundertvolt** [comment]
By the way, I just noticed your edit concerning the refresh only if a Http request is being sent.

My application triggers in mean one to three voice outputs per day, while none or more might occur as well.

My first guess would be that only if a Http request is sent at the same time (something in the order of several hours) of a token being close to expiration. At least something with a certain realistic probability of occurring by chance but for sure not always.

At least this would be a good explanation for the strongly varying time between two and seven days before the token finally "breaks".

**Comment by lsiepel** [comment]
Could you test this jar: https://1drv.ms/u/s!AnMcxmvEeupwj4hNxVqe4UVyW81Kjg?e=xTLvB9 i think it should work with 4.2.1, but it might also be 4.3.0 M1
Whenever the token is refreshed a task is scheduled that refreshed the token when it is at 80% of its lifetime. It also has some logging, so please keep the logging in debug mode.

**Comment by hundertvolt** [comment]
Great, thanks! 

I removed the standard binding and copied the jar - no issues so far installing it under OH 4.2.1. After saving a fresh token it works, and I can see "bundle org.openhab.voice.googletts:**4.3.0**.202409242114" in the logs, so it looks to me everything is up and running so far.

Please allow me one more question: If you schedule a refresh task, what happens if the system is rebooted within the wait time (which for sure happens from time to time), and what would happen if by any chance the system would be (shortly) down just at the time the task is scheduled to run (sounds like a rare case, but I've seen such things happen quite some times ;-) )?

I don't know how the token API of Google works, but at least I could imagine if there is a way to query the remaining lifetime of a token regularly (meaning once every hour or so), and if remaining lifetime is less than some sensible value (6 hours...12 hours?), then immediately refresh. This would be independent from any schedule and work through any reboot.
The only case which would lead to expiration would be an extended downtime of the system, but that's inevitable anyway.


Here comes the logging output for initialization (i.e. saving a new, valid token).
Please note that it seems to me it ran through the init process **twice** - I do not remember having seen this effect before.

```
2024-09-25 09:06:58.849 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : ConfigurableComponentHolder configuration updated for pid org.openhab.voice.googletts with change count 43

2024-09-25 09:06:58.850 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Querying state active

2024-09-25 09:06:58.851 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Querying state active

2024-09-25 09:06:58.852 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : getting modified: updateConfig

2024-09-25 09:06:58.853 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Locating method updateConfig in class org.openhab.voice.googletts.internal.GoogleTTSService

2024-09-25 09:06:58.854 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Declared Method org.openhab.voice.googletts.internal.GoogleTTSService.updateConfig([interface org.osgi.service.component.ComponentContext]) not found

2024-09-25 09:06:58.855 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Found modified method: private void org.openhab.voice.googletts.internal.GoogleTTSService.updateConfig(java.util.Map)

2024-09-25 09:06:58.856 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : invoking modified: updateConfig: parameters [org.apache.felix.scr.impl.helper.ReadOnlyDictionary]

2024-09-25 09:06:58.856 [DEBUG] [.googletts.internal.GoogleTTSService] - Updating configuration

2024-09-25 09:06:58.857 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Trying to get access and refresh tokens.

2024-09-25 09:06:59.177 [TRACE] [.googletts.internal.GoogleTTSService] - Initializing voices

2024-09-25 09:06:59.178 [TRACE] [.googletts.internal.GoogleTTSService] - Google Cloud TTS voice: gu-IN-Wavenet-B

2024-09-25 09:06:59.178 [TRACE] [.googletts.internal.GoogleTTSService] - Google Cloud TTS voice: sv-SE-Standard-D

... (Huge amount of voice inits) ...

2024-09-25 09:06:59.562 [TRACE] [.googletts.internal.GoogleTTSService] - Google Cloud TTS voice: kn-IN-Wavenet-D

2024-09-25 09:06:59.563 [TRACE] [.googletts.internal.GoogleTTSService] - Initializing audio formats

2024-09-25 09:06:59.565 [WARN ] [.googletts.internal.GoogleTTSService] - Audio format OGG_OPUS is not yet supported.

2024-09-25 09:06:59.566 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format not supported: OGG_OPUS

2024-09-25 09:06:59.567 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format supported: MP3

2024-09-25 09:06:59.568 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format supported: LINEAR16

2024-09-25 09:06:59.568 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : invoked modified: updateConfig

2024-09-25 09:06:59.569 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency $000: currently registered: true

2024-09-25 09:06:59.570 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency $001: currently registered: true

2024-09-25 09:06:59.570 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency $002: currently registered: true

2024-09-25 09:06:59.571 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency osgi.ds.satisfying.condition: currently registered: true

2024-09-25 09:06:59.572 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Querying state active

2024-09-25 09:06:59.577 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : ImmediateComponentHolder Finished configuring the dependency managers for component for pid org.openhab.voice.googletts 

2024-09-25 09:06:59.577 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : ImmediateComponentHolder Will not enable component for pid org.openhab.voice.googletts: holder enabled state: true, metadata enabled: true 

2024-09-25 09:06:59.593 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : ConfigurableComponentHolder configuration updated for pid org.openhab.voice.googletts with change count 44

2024-09-25 09:06:59.593 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Querying state active

2024-09-25 09:06:59.594 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Querying state active

2024-09-25 09:06:59.595 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : invoking modified: updateConfig: parameters [org.apache.felix.scr.impl.helper.ReadOnlyDictionary]

2024-09-25 09:06:59.596 [DEBUG] [.googletts.internal.GoogleTTSService] - Updating configuration

2024-09-25 09:06:59.671 [TRACE] [.googletts.internal.GoogleTTSService] - Initializing voices

2024-09-25 09:06:59.673 [TRACE] [.googletts.internal.GoogleTTSService] - Google Cloud TTS voice: nl-BE-Standard-A

2024-09-25 09:06:59.674 [TRACE] [.googletts.internal.GoogleTTSService] - Google Cloud TTS voice: pa-IN-Wavenet-C

... (yet another huge amount of initialized voices) ...

2024-09-25 09:07:00.040 [TRACE] [.googletts.internal.GoogleTTSService] - Google Cloud TTS voice: en-US-Standard-G

2024-09-25 09:07:00.041 [TRACE] [.googletts.internal.GoogleTTSService] - Initializing audio formats

2024-09-25 09:07:00.042 [WARN ] [.googletts.internal.GoogleTTSService] - Audio format OGG_OPUS is not yet supported.

2024-09-25 09:07:00.043 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format not supported: OGG_OPUS

2024-09-25 09:07:00.045 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format supported: MP3

2024-09-25 09:07:00.047 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format supported: LINEAR16

2024-09-25 09:07:00.048 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : invoked modified: updateConfig

2024-09-25 09:07:00.049 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency $000: currently registered: true

2024-09-25 09:07:00.050 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency $001: currently registered: true

2024-09-25 09:07:00.051 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency $002: currently registered: true

2024-09-25 09:07:00.053 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : No change in target property for dependency osgi.ds.satisfying.condition: currently registered: true

2024-09-25 09:07:00.054 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : Querying state active

2024-09-25 09:07:00.059 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : ImmediateComponentHolder Finished configuring the dependency managers for component for pid org.openhab.voice.googletts 

2024-09-25 09:07:00.060 [DEBUG] [.googletts.internal.GoogleTTSService] - bundle org.openhab.voice.googletts:4.3.0.202409242114 (336)[org.openhab.voice.googletts.internal.GoogleTTSService(438)] : ImmediateComponentHolder Will not enable component for pid org.openhab.voice.googletts: holder enabled state: true, metadata enabled: true 
```

**Comment by hundertvolt** [comment]
Hi again, 
meanwhile, things started happening. About 1:20h after the initial token, I see very frequent token refreshing actions, which started with the first actual voice output.

The refresh rate of about one hour seems unreasonably high to me. Is it really expiring so fast (and if so, why did it work for days before expiring without your fixes)  - or did you set this on purpose for not having to wait for days until the first debug log? ;-)

```
2024-09-25 10:27:40.651 [DEBUG] [.googletts.internal.GoogleTTSService] - Synthesize ' ... ' for voice 'googletts:deDEStandardE' in format AudioFormat [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100, channels=1]

2024-09-25 10:27:40.800 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 11:15:39.802 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 11:15:39.932 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 12:03:38.933 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 12:03:39.060 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 12:51:38.061 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 12:51:38.280 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 13:39:37.282 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 13:39:37.413 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 14:27:36.414 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 14:27:36.573 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 15:15:35.574 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 15:15:35.694 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 16:03:34.695 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 16:03:34.821 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 16:51:33.822 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 16:51:33.995 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 17:39:32.995 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 17:39:33.155 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 17:44:03.890 [DEBUG] [.googletts.internal.GoogleTTSService] - Synthesize ' ... ' for voice 'googletts:deDEStandardE' in format AudioFormat [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100, channels=1]

2024-09-25 18:27:32.156 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 18:27:32.354 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 19:15:31.355 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 19:15:31.476 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 19:20:59.843 [DEBUG] [.googletts.internal.GoogleTTSService] - Synthesize ' ... ' for voice 'googletts:deDEStandardE' in format AudioFormat [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100, channels=1]

2024-09-25 20:03:30.477 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 20:03:30.611 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 20:51:29.612 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 20:51:29.739 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599

2024-09-25 21:39:28.740 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Forcing a token refresh, because it is about to expire.

2024-09-25 21:39:28.865 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Auth Token Refreshed, expires in 3599
```

**Comment by hundertvolt** [comment]
Hi,
this morning, I checked the logs again. The scheme mentioned before continued so far, but then ended with this error:

`2024-09-26 06:28:04.113 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Failed to refresh the token by force. Exception in oauth communication, grant type refresh_token`

Then, even some time before, starting at around 5:30 in the morning, something extremely strange started happening.
Piece by piece over one or two hours, OpenHAB lost connection to all external devices, meanwhile all bridges are marked offline, although they are all up and running.

I have never experienced or read about such thing before, there were no useful hints in the logfile (just "status changed"), processor load and memory looked fine.

Therefore, I have no real clue if this some side effect of the experimental binding, as it happened shortly after installation, or it might be something completely different which by pure chance happened right now for the first time. But for stability, I now deleted the jar file, installed the regular binding, rebooted the system and will monitor how it's going to behave...

**Comment by lsiepel** [comment]
Thanks for sharing these details, they are usefull. 

Appearently the token expires every hour, this was already the case and is set by the remote service. I forced a refresh at 80% of its lifecycle. and that is what you see in the log. So i can confirm the mechanism works as expected. Due to this change the interval is shortened a bit, but it is also prevented to expire. I can increase it to 95% but that won't change much.

The issue you mention regarding the failed token refresh, did it recover from that or was it a fatal situation (can you provide some more logs before and after that? 

Based on your comments about devices going offline and the oath excpetion i conclude there was some transient network issue. This might be the root cause of the  token issue, it might also be a coincedence.  Hard to say, it needs more testing. I can't see any relation between this test binding and these observations regarding other Things going offline. I would propose that you continue this testing to see what happens in the next few days.

**Comment by lsiepel** [comment]
> there is already some refreshing algorithm in the code which basically does its job, but has glitches from time to time (which would be my guess).

I also think this is the case.

**Comment by wborn** [comment]
Did you also check if your Google OAuth consent screen is set to production?

If not tokens also stop working, see:

https://community.openhab.org/t/authorization-code-must-be-renewed-every-few-days/122007?u=wborn

**Comment by hundertvolt1** [comment]
> Did you also check if your Google OAuth consent screen is set to production?

OMG No, it was set to "Test"... Meanwhile I set it to Production.

I would assume a very high probability that this would make the original binding work as intended - I will try it out as soon as possible.

The interesting question is - can this information be found in the docs? Or easily by searching? 

Because initially, about more than half a year ago, I went through the step-by-step setup, everything went well, but I did not trip over that piece of info. And in the forum it seems that many others did not see or find it as well.

**Comment by lsiepel** [comment]
Let's check if that is the cause, and adapt the documentation if that seems to be the issue. Besides a network issue you had, it doesn't look like the binding is actually doing anything wrong.

**Comment by wborn** [comment]
> can this information be found in the docs? 

To set it to "production" is already part of the [configuration steps](https://www.openhab.org/addons/voice/googletts/#configuration).

>  1. Go to `OAuth consent screen` ([link](https://console.cloud.google.com/apis/credentials/consent)) and set the `publishing status` to `production` to activate the created project.

**Comment by hundertvolt** [comment]
> Let's check if that is the cause
I switch back to the original binding and let it run, if it stays stable for > 1 week I would assume that this was the cause and will let you know

> Besides a network issue you had, it doesn't look like the binding is actually doing anything wrong
As of now (your binding running stable for two and a half days), I can confirm this. 
The trouble was no network issue, but something local on the RPI - some Raspbian update trouble I suppose meanwhile, there were some updates shortly before the binding test, and I also found the Samba service completely halted. But I would not worry about the binding anymore.

> To set it to "production" is already part of the [configuration steps](https://www.openhab.org/addons/voice/googletts/#configuration).

True. But as I was definitely not the only one who had this, it seems it's a) easily to overlook (yup, my fault probably ^^) and b) there is no hint that missing this step leads to the kind of auth issues. So if you miss the step and start troubleshooting and searching, it seems to be hard to be pointed to this as the root cause.

--> maybe mention this in the manual, or add a troubleshooting hint?

Thanks so far, I'll keep you updated if it works now! :-)

**Comment by hundertvolt** [comment]
Hi again,

after one week of operation, I could find no flaws anymore after updating the Google OAuth to "Production", so I suppose that was the cause.

I was curious if I _**really**_ overlooked that one line in the setup manual, as I followed it thoroughly at that time, so I checked the file's history in GIT.

Well.... Last year, there was an update to `README.md`. This update included the line with the "production" setting, and there was no such hint there before. And guess what. The update was committed only **some few days after** my setup. And of course I did not look into the manual again, as I supposed a fault in the binding's cert renewal, as it actually worked for some days / manually. Argh. :-P  

But very good to know, as there are some more users out there suffering the same ;-) I'll update this in the forum within the next days as a solution. 

Very much thanks to all of you for your support, I'll close the ticket now.
